File: deck/seed/code/native/node/file/asynchronous/runtime/stat.ts
// File metadata for node, over `fs.promises.stat`. It builds the emitted `FileMetadata` record directly (the shim
// is prepended to the module that declares it, so the type is in scope), which is what keeps one stat from
// becoming seven. Reached only through the public file/metadata API.
import * as fs from 'node:fs/promises'

const stat = {
  metaRead: async (
    at: string,
    follow: boolean,
  ): Promise<{
    size: number
    kind: string
    made: number
    changed: number
    opened: number
    mode: number
    link: boolean
  }> => {
    try {
      const stats = follow ? await fs.stat(at) : await fs.lstat(at)
      const kind = stats.isDirectory()
        ? 'directory'
        : stats.isSymbolicLink()
          ? 'link'
          : stats.isFile()
            ? 'file'
            : 'other'

      return {
        size: stats.size,
        kind,
        made: Math.trunc(stats.birthtimeMs),
        changed: Math.trunc(stats.mtimeMs),
        opened: Math.trunc(stats.atimeMs),
        mode: stats.mode,
        link: stats.isSymbolicLink(),
      }
    } catch {
      // a missing path reads as the zero record rather than throwing: the public API is total
      return {
        size: 0,
        kind: 'other',
        made: 0,
        changed: 0,
        opened: 0,
        mode: 0,
        link: false,
      }
    }
  },
}

File: deck/seed/code/native/swift/file/asynchronous/runtime/stat.swift
// File metadata for the swift target, over NIOFileSystem's `info`. It builds the emitted `FileMetadata` record
// directly (the shim is prepended to the module that declares it, so the type is in scope), which is what keeps
// one stat from becoming seven. Reached only through the public file/metadata API.
import Foundation
import _NIOFileSystem

enum stat {
  // milliseconds since the epoch, the unit every backend's metadata reports
  private static func moment(_ time: FileInfo.Timespec) -> Int {
    time.seconds * 1000 + time.nanoseconds / 1_000_000
  }

  static func metaRead(_ path: String, _ follow: Bool) async -> FileMetadata {
    let info = try? await FileSystem.shared.info(
      forFileAt: FilePath(path),
      infoAboutSymbolicLink: !follow
    )

    // a missing path reads as the zero record rather than throwing: the public API is total
    guard let info else {
      return FileMetadata(
        size: 0,
        kind: "other",
        made: 0,
        changed: 0,
        opened: 0,
        mode: 0,
        link: false
      )
    }

    let kind: String
    switch info.type {
    case .directory: kind = "directory"
    case .symlink: kind = "link"
    case .regular: kind = "file"
    default: kind = "other"
    }

    // POSIX has no birth time in `stat`, and NIOFileSystem does not surface the darwin one: the last status
    // change is the closest honest answer, and it is what `made` means on this backend.
    return FileMetadata(
      size: Int(info.size),
      kind: kind,
      made: moment(info.lastStatusChangeTime),
      changed: moment(info.lastDataModificationTime),
      opened: moment(info.lastAccessTime),
      mode: Int(info.permissions.rawValue),
      link: info.type == .symlink
    )
  }
}
